lang/funcs/collection.go

```diff
@@ -106,6 +107,145 @@ var LengthFunc = function.New(&function.Spec{
 	},
 })
 
+// CoalesceListFunc contructs a function that takes any number of list arguments
+// and returns the first one that isn't empty.
+var CoalesceListFunc = function.New(&function.Spec{
+	Params: []function.Parameter{},
+	VarParam: &function.Parameter{
+		Name:             "vals",
+		Type:             cty.DynamicPseudoType,
+		AllowUnknown:     true,
+		AllowDynamicType: true,
+		AllowNull:        true,
+	},
+	Type: func(args []cty.Value) (ret cty.Type, err error) {
+		if len(args) == 0 {
+			return cty.NilType, fmt.Errorf("at least one argument is required")
+		}
+
+		argTypes := make([]cty.Type, len(args))
+
+		for i, arg := range args {
+			argTypes[i] = arg.Type()
+		}
+
+		retType, _ := convert.UnifyUnsafe(argTypes)
+		if retType == cty.NilType {
+			return cty.NilType, fmt.Errorf("all arguments must have the same type")
+		}
+
+		return retType, nil
+	},
+	Impl: func(args []cty.Value, retType cty.Type) (ret cty.Value, err error) {
+
+		vals := make([]cty.Value, 0, len(args))
+		for _, arg := range args {
+
+			// We already know this will succeed because of the checks in our Type func above
+			arg, _ = convert.Convert(arg, retType)
+
+			it := arg.ElementIterator()
+			for it.Next() {
+				_, v := it.Element()
+				vals = append(vals, v)
+			}
+
+			if len(vals) > 0 {
+				return cty.ListVal(vals), nil
+			}
+		}
+
+		return cty.NilVal, fmt.Errorf("no non-null arguments")
+	},
+})
+
+// CompactFunc contructs a function that takes a list of strings and returns a new list
+// with any empty string elements removed.
+var CompactFunc = function.New(&function.Spec{
+	Params: []function.Parameter{
+		{
+			Name: "list",
+			Type: cty.List(cty.String),
+		},
+	},
+	Type: function.StaticReturnType(cty.List(cty.String)),
+	Impl: func(args []cty.Value, retType cty.Type) (ret cty.Value, err error) {
+
+		var outputList []cty.Value
+
+		for it := args[0].ElementIterator(); it.Next(); {
+			_, v := it.Element()
+			if v.AsString() == "" {
+				continue
+			}
+			outputList = append(outputList, v)
+		}
+		return cty.ListVal(outputList), nil
```

Comment:
There's a gotcha here: `cty.ListVal` infers the element type of the list by analyzing the given elements (in `outputList`, here), but it can't do that if the list is empty, so there's a special constructor for empty lists:

```go
if len(outputList) == 0 {
    return cty.ListValEmpty(cty.String), nil
}
return cty.ListVal(outputList), nil
```

After seeing this, I also realize that there isn't a test case for either an empty list or a list that contains only empty strings. I think you'll see that if you add those you'll get a panic on this line. (This special case didn't apply to the old implementation, because HIL doesn't track the element types of lists.)